sourcerer-core/src/main/java/org/elder/sourcerer2/subscription/SubscriptionWorker.kt

```diff
@@ -0,0 +1,140 @@
+package org.elder.sourcerer2.subscription
+
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.channels.ClosedReceiveChannelException
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.consume
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.yield
+import org.elder.sourcerer2.EventRecord
+import org.elder.sourcerer2.EventRepository
+import org.elder.sourcerer2.EventSubscriptionHandler
+import org.elder.sourcerer2.EventSubscriptionPositionSource
+import org.elder.sourcerer2.EventSubscriptionUpdate
+import org.elder.sourcerer2.SubscriptionWorkerConfig
+import org.slf4j.LoggerFactory
+
+internal class SubscriptionWorker<T>(
+        private val repository: EventRepository<T>,
+        private val shard: Int?,
+        private val positionSource: EventSubscriptionPositionSource,
+        private val handler: EventSubscriptionHandler<T>,
+        private val config: SubscriptionWorkerConfig
+) {
+    private var retryCount = 0
+
+    suspend fun run() {
+        while (true) {
+            try {
+                logger.debug("Subscription session starting")
+                runOneSession()
+                // We're still here meaning the subscription marked completion, clean exit
+            } catch (ex: Exception) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Caught exception is too generic. Prefer catching specific Exceptions to handle specific error cases. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3ACatchException)